lib/modules/manager/nix/artifacts.spec.ts

```diff
@@ -0,0 +1,241 @@
+import { GlobalConfig } from '../../../config/global';
+import type { RepoGlobalConfig } from '../../../config/types';
+import * as docker from '../../../util/exec/docker';
+import type { UpdateArtifactsConfig } from '../types';
+import { updateArtifacts } from '.';
+
+jest.mock('../../../util/exec/env');
+jest.mock('../../../util/fs');
+jest.mock('../../../util/git');
+
+const adminConfig: RepoGlobalConfig = {
+  // `join` fixes Windows CI
+  localDir: join('/tmp/github/some/repo'),
+  cacheDir: join('/tmp/renovate/cache'),
+  containerbaseDir: join('/tmp/renovate/cache/containerbase'),
+};
+const dockerAdminConfig = { ...adminConfig, binarySource: 'docker' };
+
+process.env.BUILDPACK = 'true';
+
+const config: UpdateArtifactsConfig = {};
+const lockMaintenanceConfig = { ...config, isLockFileMaintenance: true };
+const updateInputCmd = `nix \
+    --extra-experimental-features nix-command \
+    --extra-experimental-features flakes \
+    flake lock --update-input nixpkgs`;
+const lockfileMaintenanceCmd = `nix \
+    --extra-experimental-features nix-command \
+    --extra-experimental-features flakes \
+    flake update`;
+
+describe('modules/manager/nix/artifacts', () => {
+  beforeEach(() => {
+    jest.resetAllMocks();
+    env.getChildProcessEnv.mockReturnValue({
+      ...envMock.basic,
+      LANG: 'en_US.UTF-8',
+      LC_ALL: 'en_US',
+    });
+    GlobalConfig.set(adminConfig);
+    docker.resetPrefetchedImages();
+  });
+
+  it('returns if no flake.lock found', async () => {
+    const execSnapshots = mockExecAll();
+    expect(
+      await updateArtifacts({
+        packageFileName: 'flake.nix',
+        updatedDeps: [],
+        newPackageFileContent: '',
+        config,
+      })
+    ).toBeNull();
+    expect(execSnapshots).toEqual([]);
+  });
+
+  it('returns null if unchanged', async () => {
+    fs.readLocalFile.mockResolvedValueOnce('content');
+    const execSnapshots = mockExecAll();
+    git.getRepoStatus.mockResolvedValue({
+      modified: [''],
+    } as StatusResult);
+
+    expect(
+      await updateArtifacts({
+        packageFileName: 'flake.nix',
+        updatedDeps: [{ depName: 'nixpkgs' }],
+        newPackageFileContent: 'some new content',
+        config,
+      })
+    ).toBeNull();
+    expect(execSnapshots).toMatchObject([{ cmd: updateInputCmd }]);
+  });
+
+  it('returns updated flake.lock', async () => {
+    fs.readLocalFile.mockResolvedValueOnce('current flake.lock');
+    const execSnapshots = mockExecAll();
+    git.getRepoStatus.mockResolvedValue({
+      modified: ['flake.lock'],
+    } as StatusResult);
+    fs.readLocalFile.mockResolvedValueOnce('new flake.lock');
+
+    expect(
+      await updateArtifacts({
+        packageFileName: 'flake.nix',
+        updatedDeps: [{ depName: 'nixpkgs' }],
+        newPackageFileContent: 'some new content',
+        config: { ...config, constraints: { python: '3.7' } },
+      })
+    ).not.toBeNull();
```

Comment:
Use `toEqual` instead of `not.toBeNull`, otherwise we silently allow artifact errors instead of successful update.

We had that on other managers already while working on install mode.